feat: add new flag --experimental-enable-top-level-await

harmony7 opened the pull request:
This PR is a fix for #729

These fixes were enough to get top-level-await in both my main module and dependencies working for me.

I would love to add a test suite, but wasn't able to figure out how to do it. Does the setup require a linux machine like it says in the DEVELOPING readme file? I'm on a Mac...

The following minimal case works (requires `"type": "module"` in `package.json`):

src/index.js
```javascript
/// <reference types="@fastly/js-compute" />

import { foo } from './dep.js';
const value = await foo();

addEventListener("fetch", (event) => event.respondWith(handleRequest(event)));
async function handleRequest(event) {
  return new Response(value, { status: 200 }); // returns 'bar'
}
```

src/dep.js
```javascript
export async function wait() {
  return Promise.resolve();
}

await wait();

export async function foo() {
  await wait();
  return 'bar';
}
```

These changes also enabled me to use top level await in my upcoming version of compute-js-opentelemetry that no longer uses Webpack.

I've also tested and seen that `"type": "commonjs"` in `package.json` still works.

harmony7 commented:
Also, WRT to error handling, it seems that with this implementation, if the top level await throws an exception, this happens:

```javascript
export async function foo() {
  throw new Error('foo');
}
```

```
Promise rejected but never handled: foo
Stack:
  foo@<stdin>:7:9
  @<stdin>:13:19
  async*@<stdin>:18:3
```

If more robust handling is required, please advise.

JakeChampion commented on src/bundle.js:
```diff
@@ -68,6 +68,7 @@ export async function bundle(input) {
     entryPoints: [input],
     bundle: true,
     write: false,
+    format: 'esm',
```
From [esbuild docs](https://esbuild.github.io/api/#format-esm):
>The esm format stands for "ECMAScript module". It assumes the environment supports import and export syntax.

We don't support import and export syntax, do you know if any issues will arise from this change?

harmony7 commented:
I've tried putting it behind a flag in its own changeset

harmony7 commented:
I've pushed a little more code.

The first changeset enables TLA in two steps:

* `bundle()` uses `format: 'esm'` when calling esbuild (also updated `precompile()` to use `sourceType: 'module'`)
* Added `enableTopLevelAwait()` that:
  * removes export declarations from the bundle (taking care to leave the declaration part intact for exports that have them)
  * wraps the entire thing in an async iife

In the second changeset:
All this is only enabled if `--experimental-enable-top-level-await` is specified. Otherwise the behavior is identical to the prior behavior.
